programmazione funzionale + risoluzione bug impaginazione

francescogemma opened the pull request:
Stesura del capitolo sulla programmazione funzionale, corretto un bug dell'impaginazione nell'esporta.js e aggiunta la parte di Style.css che permette ad una singola slide di essere a due colonne (mentre le altre rimangono a 3 colonne)

francescogemma committed:
programmazione funzionale + risoluzione bug impaginazione

Stesura del capitolo sulla programmazione funzionale e corretto un bug dell'impaginazione nell'esporta.js

m1gwings commented on esporta.js:
```diff
@@ -48,9 +48,11 @@ for (let i = 0; i < indice.capitoli.length; i++) {
+    markdown += markdownCapitolo + ((i < indice.capitoli.length - 1) ? '\n\n---\n\n' : '')
 }
 
+// console.log(markdown)
```
Potresti rimuovere il commento in `esporta.js`?

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
 # Programmazione funzionale
+
+Il concetto di programmazione funzionale si basa sul fatto di considerare le funzioni come valori, e quindi utilizzarle come tali. In questo modo, le funzioni possono essere passate come parametri, ritornate come risultati, memorizzate in variabili, ecc.
```
Cambierei la frase iniziale con quella che segue:
_Il paradigma di programmazione funzionale si basa sul principio di considerare le funzioni come oggetti_

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+### `Filter`
+
+Possiamo filtrare lo `stream` per far si che tutti gli elementi della collezione soddisfino un cetro predicato (funzione che ritorna solo true o false).
```
Se spieghi cos'è il predicato prima, credo si possa rimuovare la parte tra parentesi

francescogemma replied:
Non l'ho spiegato prima, credo le l'elenco Predicato, Funzione, ... (quello che c'è dalle slide) sia scontato o per lo meno scomodo all'esame (in cui devi avere tutto sottomano subito). Quindi l'ho spiegato "on the fly"

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+### `flatMap`
+
+**`flatMap(<funzione da T a Stream<U>>)`** prende in input una funzione, la applica ad ogni elemento di tipo T dello `stream`. La funzione ritorna un altro `stream` (in generale contentente elementi di tipo diverso U) per ogni elemento. Infine tutti gli `stream` vengono concatenati in un unico `stream`.
```
Cambierei la frase finale con:
```markdown
Infine tutti gli `stream` restituiti dalla funzione vengono concatenati in un unico `stream`. 
```

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+Otterrò uno `stream`, contenente le stringhe: "Luigi", "30", "Pippo", "40", "Pluto", "50".
+
+### `Reduce`
```
Credo vada tolta la maiuscola

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+### `Reduce`
+
+La **`reduce(Identità, <funzione binaria>)`** è un'operazione per avere un singolo valore a partire da uno `stream`. 
```
Cambierei con
```markdown
... è un'operazione di aggregazione che restituisce un singolo valore a partire da uno `stream`.
```

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+La **`reduce(Identità, <funzione binaria>)`** è un'operazione per avere un singolo valore a partire da uno `stream`. 
+
+Si ha come primo parametro un valore iniziale (identità), e come secondo parametro una funzione binaria che prende in input due elementi dello `stream` e ritorna un altro elemento (dello stesso tipo) ottenuto a partire dai due. La funzione verrà applicata ricorsivamente ad ogni coppia di elementi dello `stream`, fino ad ottenere un solo elemento.
```
Credo che sia fuorviante scrivere che la funzione viene applicata ad ogni coppia di elementi dello stream dato che viene applicata, ogni volta, ad un elemento dello stream ed al risultato parziale della riduzione calcolato al punto precedente.
Se riesci cerca di spiegare meglio e rendere più chiara questa cosa, magari anche elaborando con l'esempio sulla somma mostrando un po' le varie iterazioni, tipo:
```markdown
[30, 40, 50, 60, ...]
30 + 40 = 70 -> 70 + 50 = 120 -> 120 + 60 = 180 -> ...
```
Anche perché credo che potrebbe capitare di dover implementare un'operazione di riduzione custom ed è importante avere chiaro questo concetto "induttivo".

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+Otterrò un intero: 120 (= 30 + 40 + 50).
+
+### `Collect`
```
Credo vada tolta la maiuscola

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+        p.getNome() + ": " + 
+        p.getEta().toString()
+    );
+    .collect(Collectors.toList());
```
Se riesci spiega la faccenda dei collectors, di come si implementa per una collezione custom e del fatto che per le liste è già definito. (è una cosa secondaria, si può fare anche più in la nelle prossime iterazioni di scrittura).

francescogemma replied:
Se proprio serve, lo farei più in là. Nei temi d'esame non ho mai visto di implementare un Collector custom (ricorda che serve esclusivamente per l'esame questo documento)

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+Possiamo creare un `Optional` con il valore `val` con il metodo `Optional.of(val)`. Oppure con il metodo `Optional.empty()` per creare un `Optional` vuoto (sostituisce il null).
+
+Sono poi disponibili i metodi per lavorare con gli Optional:
+
```
Cerca di elaborare un po' di più su come si utilizzano gli optional, magari puoi fare anche qualche esempio

francescogemma replied:
Ho aggiunto un esempio

m1gwings commented on programmazione-funzionale.md:
```diff
@@ -1 +1,198 @@
+
+- **`orElse(<val>)`** che ritorna il valore dell'Optional se non è vuoto, altrimenti ritorna il valore `val` in input.
+
+- **`flatMap(<funzione da T a Optional<U>)`** che ritorna un `Optional<U>`, se è vuoto ritorna vuoto, altrimenti applica la funzione.
```
Credo vada spiegato un po' meglio come funziona flatMap, o almeno, io non ho ancora ripetuto come funziona (ed al momento non lo ricordo) e leggendo la spiegazione mi risulta poco chiaro.

m1gwings reviewed (changes requested):
Ottimo lavoro! Ho lasciato qualche commento per riorganizzare un po' la struttura del capitolo ed aggiungere qualche contenuto secondo me importante (ovviamente se ne può discutere).
Oltre alle modifiche che ho specificato, se riesci, quando introduci una funzione di Java, oltre al nome ed ai parametri, puoi aggiungere il valore restituito, o prima della funzione
```markdown
<valore-restituito> nomeFunzione(<...>)
```
o con una freccetta
```markdown
nomeFunzione(<...>) -> <valore-restiuito>
```

Perchè ad esempio guardando i metodi delle stream (tipo flatMap) non mi è chiaro, considerando solo la signature, cosa restituiscono.

Infine se riesci, sostituisci "ritorna" con "restituisce" (per la maggior parte dei casi basta fare un replace all, agli altri posso anche pensarci io manualmente).

francescogemma committed:
aggiornato capitolo programmazione-funzionale

Ho aggiornato il capitolo della programmazione funzionale seguendo le linee guida

francescogemma committed:
Update programmazione-funzionale.md

francescogemma commented:
ho solo pushato, dovrebbe averlo aggiunto alla pull request

francescogemma committed:
aggiunto esercizio da tema d'esame

aggiunto esercizio da tema d'esame in fondo al capitolo sulla programmazione funzionale
